[HigherOrderOp] set set_subgraph_inputs to flatten_manual for map

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115853

We change manually_set_subgraph_inputs to three modes: manual, automatic and flatten_manual. The flatten_manual wil first flatten the sub_args then recussively call set_subgrah_inputs = "manual". This allows us to control the order of the placeholder shown up in the graph, which is necessary for map, where we want to keep the mapped arguments before the rest positional arguments.

Right now, map only takes a single tensor as mapped argument but it would become pretty easy to match the subgraph inputs to original proxy if we have a "flatten_manual" option.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ydwu4 committed:
[HigherOrderOp] set manually_set_subgraph_inputs to False for map

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -746,9 +746,67 @@ def call_function(
+        # causes their parent to lift to their graph inputs (they're not used anywhere in the parent
+        # since it's a temporary slice from parent graph's point of view) and
+        # finally becomes a get_item call in top-level grah.
+        used_lifted_freevars = {
```
Talked to @zou3519  offline. This gluing code is hard to maintain. Will try the flatten-unflatten approach + dce to avoid gluing code like this.

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -746,7 +771,9 @@ def call_function(
             source_target=self.value,
+            manually_set_subgraph_inputs=False,
             should_flatten_outputs=True,
+            should_flatten_inputs=True,
```
make it one option of mannually_set_subgraph_inputs

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"


We change manually_set_subgraph_inputs to three modes: manual, automatic and flatten_manual. The flatten_manual wil first flatten the sub_args then recussively call set_subgrah_inputs = "manual". This allows us to control the order of the placeholder shown up in the graph, which is necessary for map, where we want to keep the mapped arguments before the rest positional arguments.

Right now, map only takes a single tensor as mapped argument but it would become pretty easy to match the subgraph inputs to original proxy if we have a "flatten_manual" option.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -196,14 +213,16 @@ def speculate_subgraph(
-    # If manually_set_subgraph_inputs=True, then we manually add
-    # the `sub_args` to `subgraph`, if False then we rely
-    # on tracer's lifting mechanism to lift these args.
+    # NOTE [Temporary argument `set_subgraph_inputs`]
+    # If set_subgraph_inputs="manual", we manually add the `sub_args` to `subgraph`,
+    # If "automatic", we rely on tracer's lifting mechanism to lift these args.
+    # If "flatten_manual", we first flatten the sub_args into a flattend list, then manually
+    # set the flattend inputs to the subgraph. This saves us from having to mapping the automatically
+    # lifted inputs back to their original position, where ordering of input matters e.g. map.
```
This arg isn't temporary anymore. Let's delete the next 3 lines ("Default `True` is temporary").


We should say something like
- if your HOP supports arbitrary inputs, use set_subtraph_inputs="automatic"
- if your HOP supports Tensor and symnode inputs, use set_subgraph_inputs="flatten_manual"

ydwu4 replied:
For "manual" option, the arg can actually be AutogradFunctionContextVariable or it supports maybe_fx_node(), maybe we can change the recommendation to "This is the most restrictive option (your function must only accept Tensor and symnode inputs or you know exactly how to create a proxy for it) and we do not recommend it."

zou3519 replied:
Yeah go for it

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -196,14 +213,16 @@ def speculate_subgraph(
     #       to always lift args in future and remove this
     #       argument.
-    manually_set_subgraph_inputs=True,
+    set_subgraph_inputs="manual",
```
Actually, do we need "manual" anymore? "flatten_manual" seems like it is better than "manual"

zou3519 replied:
Nvm, we still need manual if preserving the calling convention matters to people.

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -759,11 +778,11 @@ def call_function(
             1,  # right now we only supports num_mapped = 1
-            *(arg.as_proxy() for arg in args[1:]),
-            *(arg for arg in body_lifted_freevars.keys()),
+            *([arg.as_proxy() for arg in args[1:]] + list(body_lifted_freevars.keys())),
```
Is this correct? I would have expected that we need to flatten args first

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"


We change manually_set_subgraph_inputs to three modes: manual, automatic and flatten_manual. The flatten_manual wil first flatten the sub_args then recussively call set_subgrah_inputs = "manual". This allows us to control the order of the placeholder shown up in the graph, which is necessary for map, where we want to keep the mapped arguments before the rest positional arguments.

Right now, map only takes a single tensor as mapped argument but it would become pretty easy to match the subgraph inputs to original proxy if we have a "flatten_manual" option in the following diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"


We change manually_set_subgraph_inputs to three modes: manual, automatic and flatten_manual. The flatten_manual wil first flatten the sub_args then recussively call set_subgrah_inputs = "manual". This allows us to control the order of the placeholder shown up in the graph, which is necessary for map, where we want to keep the mapped arguments before the rest positional arguments.

Right now, map only takes a single tensor as mapped argument but it would become pretty easy to match the subgraph inputs to original proxy if we have a "flatten_manual" option in the following diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"


We change manually_set_subgraph_inputs to three modes: manual, automatic and flatten_manual. The flatten_manual wil first flatten the sub_args then recussively call set_subgrah_inputs = "manual". This allows us to control the order of the placeholder shown up in the graph, which is necessary for map, where we want to keep the mapped arguments before the rest positional arguments.

Right now, map only takes a single tensor as mapped argument but it would become pretty easy to match the subgraph inputs to original proxy if we have a "flatten_manual" option in the following diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"


We change manually_set_subgraph_inputs to three modes: manual, automatic and flatten_manual. The flatten_manual wil first flatten the sub_args then recussively call set_subgrah_inputs = "manual". This allows us to control the order of the placeholder shown up in the graph, which is necessary for map, where we want to keep the mapped arguments before the rest positional arguments.

Right now, map only takes a single tensor as mapped argument but it would become pretty easy to match the subgraph inputs to original proxy if we have a "flatten_manual" option in the following diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] set manually_set_subgraph_inputs to False for map"


We change manually_set_subgraph_inputs to three modes: manual, automatic and flatten_manual. The flatten_manual wil first flatten the sub_args then recussively call set_subgrah_inputs = "manual". This allows us to control the order of the placeholder shown up in the graph, which is necessary for map, where we want to keep the mapped arguments before the rest positional arguments.

Right now, map only takes a single tensor as mapped argument but it would become pretty easy to match the subgraph inputs to original proxy if we have a "flatten_manual" option in the following diff.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
